### PR TITLE
[AIRFLOW-987] pass kerberos cli args keytab and principal to kerberos…

### PR DESCRIPTION
**Pull request by xnuinside**
….run()

Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-987]\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-987
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

default=conf.get('kerberos', 'principal')) - removed from cli, because of behavior inside renew_from_kt function. 

inside renew_from_kt we parse conf args one way with one replace type, and if we decide what need to use workaround we parse conf args in perform_krb181_workaround() using the second way

so, it's not possible to put as default conf value in principal, because this way will be always args.principal and we ca not have 2 different ways how to parse this value

also b"\n".join(subp.stdout.readlines()) caused error on 3.5, it was not catched because wasn't covered with tests

error in 3.5 - 
` b"\n".join(subp.stderr.readlines())))

   TypeError: sequence item 0: expected a bytes-like object, str found`
### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by xnuinside** [comment]
@Fokko, @ashb please review

File: airflow/security/kerberos.py [hunk 1]
```diff
@@ -55,8 +57,8 @@ def renew_from_kt():
     if subp.returncode != 0:
         log.error("Couldn't reinit from keytab! `kinit' exited with %s.\n%s\n%s" % (
             subp.returncode,
-            b"\n".join(subp.stdout.readlines()),
-            b"\n".join(subp.stderr.readlines())))
+            "\n".join(subp.stdout.readlines()),
+            "\n".join(subp.stderr.readlines())))
```
Comment (ashb) [inline airflow/security/kerberos.py:61]: Why this change?

**Review by ashb — COMMENTED** [review]
Do we still need these lines? https://github.com/apache/incubator-airflow/blob/master/airflow/bin/airflow#L26-L28 

(I'm totally unfamiliar with Kerberos.)

File: airflow/security/kerberos.py (hunk 1 above)
Comment (xnuinside) [reply airflow/security/kerberos.py:61]: @ashb , 
also b"\n".join(subp.stdout.readlines()) caused error on 3.5, it was not catched because wasn't covered with tests

error in 3.5 -
` b"\n".join(subp.stderr.readlines())))

TypeError: sequence item 0: expected a bytes-like object, str found`

Comment (xnuinside) [reply airflow/security/kerberos.py:61]: Do we still need these lines? https://github.com/apache/incubator-airflow/blob/master/airflow/bin/airflow#L26-L28 - I don't know :/ i fix only bug, what args from cli not passed to kerberos.run, but they are defined in cli doc

File: airflow/security/kerberos.py [hunk 2]
```diff
@@ -27,19 +27,21 @@
 log = LoggingMixin().log
 
 
-def renew_from_kt():
+def renew_from_kt(args=None):
```
Comment (ashb) [inline airflow/security/kerberos.py]: I think rather than pass the whole args dict in here it would be better to pass the two we specifically want (`principal`, and `keytab`)

File: airflow/security/kerberos.py [hunk 3]
```diff
@@ -80,7 +82,7 @@ def perform_krb181_workaround():
     ret = subprocess.call(cmdv, close_fds=True)
 
     if ret != 0:
-        principal = "%s/%s" % (
+        principal = args.principal or "%s/%s" % (
```
Comment (ashb) [inline airflow/security/kerberos.py]: I'm not sure this change is right - I'd have guessed that args.principal would be used to override the principal in the config and then add to it?

File: airflow/security/kerberos.py [hunk 4]
```diff
@@ -110,11 +112,11 @@ def detect_conf_var():
         return b'X-CACHECONF:' in f.read()
 
 
-def run():
-    if configuration.conf.get('kerberos', 'keytab') is None:
+def run(args):
+    if not args.keytab:
```
Comment (ashb) [inline airflow/security/kerberos.py]: This needs an `or configuration.conf.get('kerberos', 'keytab') is None` too

File: tests/security/test_kerberos.py [hunk 5]
```diff
@@ -32,13 +33,26 @@ def setUp(self):
...
+
+    def test_args_from_cli(self):
+        """
+        We expect no result, but a run with sys.exit(1) because keytab not exist.
+        """
+        configuration.conf.set("kerberos", "keytab", "")
+        self.args.keytab = "test_keytab"
+
+        with self.assertRaises(SystemExit) as se:
+            renew_from_kt(args=self.args)
```
Comment (ashb) [inline tests/security/test_kerberos.py]: What error is this hitting? Could you add some assertion of what error is logged too? (Probably use https://docs.python.org/3.6/library/unittest.html#unittest.TestCase.assertLogs via the backported `unittest2` which is already a dependency)

**Review by ashb — CHANGES_REQUESTED** [review]

File: airflow/security/kerberos.py (hunk 4 above)
Comment (xnuinside) [reply airflow/security/kerberos.py]: no, it not needed, because we get configuration.conf.get('kerberos', 'keytab') as default value in arg - https://github.com/apache/incubator-airflow/blob/master/airflow/bin/cli.py#L1685 , so if args.keytabe will be None it mean what config empty too

File: tests/security/test_kerberos.py (hunk 5 above)
Comment (xnuinside) [reply tests/security/test_kerberos.py]: it sys.exit(1) - https://github.com/apache/incubator-airflow/blob/3dc9e4021786e2c2416956f83a1f19ea1052ff3b/airflow/security/kerberos.py#L62 about assertLogs - okay, I will add it

File: airflow/security/kerberos.py (hunk 3 above)
Comment (xnuinside) [reply airflow/security/kerberos.py]: We don't override config here. It is used only to prepare a command - https://github.com/apache/incubator-airflow/blob/3dc9e4021786e2c2416956f83a1f19ea1052ff3b/airflow/security/kerberos.py#L46

File: airflow/security/kerberos.py (hunk 2 above)
Comment (xnuinside) [reply airflow/security/kerberos.py]: okay, make sense

File: airflow/security/kerberos.py (hunk 4 above)
Comment (ashb) [reply airflow/security/kerberos.py]: Ah I missed that, thanks.

Does it come out as None or `""`? (Oh shouldn't matter for `if not`)

File: airflow/security/kerberos.py (hunk 3 above)
Comment (ashb) [reply airflow/security/kerberos.py]: This is a change in behaviour.

Previously:

```
principal == "the-actual-principal/my-fqdn"
```

After if I specify the arg:

```
principal == "the-actual-principal"
```

i.e. if arg.principal is provided the format is never called, and I think it should be?

Comment (xnuinside) [reply airflow/security/kerberos.py]: oh, I understand what you talking about, I think you right, yes. My mistake.

File: tests/security/test_kerberos.py (hunk 5 above)
Comment (xnuinside) [reply tests/security/test_kerberos.py]: @ashb New in version 3.4. method added only in 3.4, at 2.7 will fail with 
`    with self.assertLogs(LoggingMixin().log) as log:
AttributeError: 'KerberosTest' object has no attribute 'assertLogs'`

**Comment by xnuinside** [comment]
@ashb, please review

File: tests/security/test_kerberos.py (hunk 5 above)
Comment (ashb) [reply tests/security/test_kerberos.py]:
```
from unittest2 import TestCase
```

File: airflow/security/kerberos.py [hunk 6]
```diff
@@ -80,10 +82,8 @@ def perform_krb181_workaround():
     ret = subprocess.call(cmdv, close_fds=True)
 
     if ret != 0:
-        principal = "%s/%s" % (
-            configuration.conf.get('kerberos', 'principal'),
-            socket.getfqdn()
-        )
+        principal = "%s/%s" % (configuration.conf.get('kerberos', 'principal'),
```
Comment (ashb) [inline airflow/security/kerberos.py]:
```suggestion
        principal = "%s/%s" % (principal or configuration.conf.get('kerberos', 'principal'),
```

which will need the principal var passed to this function. Maybe just `principal = "%s/%s" % (principal,` would do then

Comment (xnuinside) [reply airflow/security/kerberos.py]: done

**Comment by xnuinside** [comment]
@ashb , tests passsed

**Comment by xnuinside** [comment]
@ashb ?)

**Review by ashb — APPROVED** [review]
